x-pack/filebeat/module/suricata/eve/ingest/pipeline.json

```diff
@@ -50,10 +50,22 @@
             }
         },
         {
-            "rename": {
+            "append": {
+                "if": "ctx.suricata?.eve?.http?.hostname != null",
+                "value": "{{suricata.eve.http.hostname}}",
+                "field": "destination.domain"
+            }
+        },
+        {
+            "remove": {
                 "field": "suricata.eve.http.hostname",
-                "target_field": "destination.domain",
-                "ignore_missing": true
+                "ignore_failure": true
+            }
+        },
+        {
+            "script": {
+                "source": "def domain = ctx.destination?.domain; if (domain instanceof Collection) { if (domain.length == 1) { ctx.destination.domain = domain[0]; } else { ctx.destination.domain = ctx.destination.domain.stream().distinct().collect(Collectors.toList()); } }",
+                "ignore_failure": true
```

Comment:
I wonder if domain containing an array will be a problem.

Suggestion: dedupe before `if (domain.length == 1)`.

Reply:
Yeah we'll need to start looking into defining this more clearly in ECS.

But I agree that we have to support arrays here, because of the reverse DNS possibility.

Reply:
Ah wait. You're only keeping the first of the array? 🤔How do you know it's the interesting entry?

Reply:
No. This is the code. It doesn't drop any data.

```groovy

    def domain = ctx.destination?.domain; 
    if (domain instanceof Collection) {
      // Deduplicate
      ctx.destination.domain = ctx.destination.domain.stream().distinct().collect(Collectors.toList());
      
      // Make it a plain old string if there's only one item.
      if (domain.length == 1) { 
        ctx.destination.domain = domain[0]; 
      } 
    }

```

Reply:
Can the reverse DNS processor return more than one DNS entry for a given IP?

If so, this only keeps the first one, right?

Or does the processor always return an array of one?

Reply:
Oh, I see the problem. When I refactored it from the initial version I broke it. The fixed version is:

```groovy
    def domain = ctx.destination?.domain; 
    if (domain instanceof Collection) {
      // Deduplicate
      domain = domain.stream().distinct().collect(Collectors.toList());
      
      // Make it a plain old string if there's only one item.
      if (domain.length == 1) { 
        domain = domain[0]; 
      }
      
      // Set the value
      ctx.destination.domain = domain;
    }
```

Does this address your concern?

Reply:
Ah, happy to see we weren't just going around in circles :-)

Yes, this makes sense now :-)